From cf0fb47542c8f348ecc1d0135ada6200bd82db25 Mon Sep 17 00:00:00 2001
From: Shivani Soni <quic_soni@quicinc.com>
Date: Tue, 18 Apr 2023 14:42:24 +0530
Subject: [PATCH] arm/arm64: bonding: Introduce a callback to destroy netdev
 wifi priv

Introduce a callback for bond MLO to destruct wiphy
corresponding to the bond netdev.
Handler is registered in wifi driver

Change-Id: Ied7ecc6a65dd084cc89345bc147ffc1355e8f1b2
Signed-off-by: Shivani Soni <quic_soni@quicinc.com>
---
 drivers/net/bonding/bond_main.c | 8 ++++++++
 include/net/bonding.h           | 2 ++
 2 files changed, 10 insertions(+)

diff --git a/drivers/net/bonding/bond_main.c b/drivers/net/bonding/bond_main.c
index 94cb2157e810..1e0e2683be41 100644
--- a/drivers/net/bonding/bond_main.c
+++ b/drivers/net/bonding/bond_main.c
@@ -6167,6 +6167,14 @@ static void bond_destructor(struct net_device *bond_dev)
 
 	if (bond->rr_tx_counter)
 		free_percpu(bond->rr_tx_counter);
+	/*
+	 * Wifi driver registered callback to destroy wiphy for MLO bond netdevice
+	 */
+	if (bond_is_mlo_device(bond_dev)) {
+		if (bond->mlo_info.bond_mlo_netdev_priv_destructor) {
+			bond->mlo_info.bond_mlo_netdev_priv_destructor(bond_dev);
+		}
+	}
 }
 
 void bond_setup(struct net_device *bond_dev)
diff --git a/include/net/bonding.h b/include/net/bonding.h
index 7f04f1d26194..314b60ac7fc2 100644
--- a/include/net/bonding.h
+++ b/include/net/bonding.h
@@ -220,12 +220,14 @@ struct bond_ipsec {
  * @bond_get_mlo_tx_netdev: Callback function to get link interface from wifi driver for transmit
  * @bond_mlo_ctx: Private member for wifi driver
  * @wdev: ieee80211_ptr for wifi VAP
+ * @bond_mlo_netdev_priv_destructor: Callback function to remove wiphy instance from wifi driver
  */
 struct mlo_bond_info {
 	int (*bond_mlo_xmit_netdev)(struct sk_buff *skb, struct net_device *bond_dev);
 	struct net_device *(*bond_get_mlo_tx_netdev)(void *bond_mlo_ctx, void *dst);
 	void *bond_mlo_ctx;
 	struct wireless_dev *wdev;
+	void (*bond_mlo_netdev_priv_destructor)(struct net_device *bond_dev);
 };
 
 /*
-- 
2.34.1

